### ISSUE TITLE
Failed: BalancingSpec

### ISSUE THREAD

**Issue by johanandren** [akka/akka#30359]
https://travis-ci.com/github/akka/akka/jobs/521145928

```
[info] - must deliver messages in a balancing fashion when overridden in config *** FAILED *** (359 milliseconds)
1151[info]   java.lang.AssertionError: assertion failed: received unexpected message RealMessage(1,Actor[akka://BalancingSpec/user/balancingPool-3/$a#1476506991]) after 0 millis
1152[info]   at akka.testkit.TestKitBase.expectNoMsg_internal(TestKit.scala:769)
1153[info]   at akka.testkit.TestKitBase.expectNoMessage(TestKit.scala:733)
1154[info]   at akka.testkit.TestKitBase.expectNoMessage$(TestKit.scala:159)
1155[info]   at akka.testkit.TestKit.expectNoMessage(TestKit.scala:973)
1156[info]   at akka.routing.BalancingSpec.test(BalancingSpec.scala:86)
1157[info]   at akka.routing.BalancingSpec.f$proxy3$1(BalancingSpec.scala:119)
1158[info]   at akka.routing.BalancingSpec.$init$$$anonfun$4$$anonfun$3(BalancingSpec.scala:115)
1159[info]   at org.scalatest.Transformer.apply$$anonfun$1(Transformer.scala:22)
1160
```

**Comment by raboof** [comment]
https://github.com/akka/akka/runs/3169652657?check_suite_focus=true

**Comment by ignasi35** [comment]
Again in https://github.com/akka/akka/pull/30444/checks?check_run_id=3220266807

**Comment by raboof** [comment]
merged #30438 so we might get a more useful failure message next time

**Comment by raboof** [comment]
Again in https://github.com/akka/akka/pull/30453/checks?check_run_id=3240969608 - seems like it receives the first `1` correctly, but then receives an unexpected duplicate `1` message.

The `RealMessage` wrapper *seems* like a `TestKit` internal that we should perhaps unpack and doesn't have anything to do with this test?

Perhaps we should record the sender of the first message, to make sure it's not crosstalk between the 3 tests?

**Comment by ignasi35** [comment]
Got the issue again in:


```

[info] balancing pool
[WARN] [08/05/2021 17:27:23.874] [pool-1-thread-1-ScalaTest-running-BalancingSpec] [BalancingSpec(akka://BalancingSpec)] Actor[akka://BalancingSpec/user/balancingPool-1/$a#-1119056079]
[info] - must deliver messages in a balancing fashion when defined programatically (1 second, 841 milliseconds)
[WARN] [08/05/2021 17:27:25.683] [pool-1-thread-1-ScalaTest-running-BalancingSpec] [BalancingSpec(akka://BalancingSpec)] Actor[akka://BalancingSpec/user/balancingPool-2/$b#-129009524]
[info] - must deliver messages in a balancing fashion when defined in config (1 second, 807 milliseconds)
[WARN] [08/05/2021 17:27:27.542] [pool-1-thread-1-ScalaTest-running-BalancingSpec] [BalancingSpec(akka://BalancingSpec)] Actor[akka://BalancingSpec/user/balancingPool-3/$b#343717519]
[info] - must deliver messages in a balancing fashion when overridden in config *** FAILED *** (364 milliseconds)
[info]   java.lang.AssertionError: assertion failed: received unexpected message RealMessage(1,Actor[akka://BalancingSpec/user/balancingPool-3/$b#343717519]) after 0 millis
[info]   at akka.testkit.TestKitBase.expectNoMsg_internal(TestKit.scala:769)
[info]   at akka.testkit.TestKitBase.expectNoMessage(TestKit.scala:733)
[info]   at akka.testkit.TestKitBase.expectNoMessage$(TestKit.scala:732)
[info]   at akka.testkit.TestKit.expectNoMessage(TestKit.scala:973)
```

The sender in `RealMessage(1,Actor[akka://BalancingSpec/user/balancingPool-3/$b#343717519])` is from the `overridden in config` test (see `user/balancingPool-3`). It doesn't look like cross-talk from the other test. 

It actually looks like `expectNoMessage` obtains the message received in `receiveN` above (a duplicate of the message just received):

```scala

    for (i <- 1 to iterationCount) {
      pool ! i
    }

    // all but one worker are blocked
    val replies1 = receiveN(iterationCount - poolSize + 1)
    // all replies from the unblocked worker so far
    replies1.toSet should be(Set(1))
    log.warning(lastSender.toString)
    expectNoMessage(1.second)
```

**Comment by ignasi35** [comment]
Note that this failure happens both in scala3 and scala 2.x builds. As soon as we started running this test on GH Actions this failure surfaced.

**Comment by raboof** [comment]
> Note that this failure happens both in scala3 and scala 2.x builds. As soon as we started running this test on GH Actions this failure surfaced.

As you can see at the start of this thread we've seen it once on Travis as well - but indeed it seems to be more likely to be triggered on GHA

**Comment by ignasi35** [comment]
> As you can see at the start of this thread we've seen it once on Travis as well

I need more coffee. I swear I read that link and thought: Ha! It's always on Github! 🤦🏼‍♂️

**Comment by ignasi35** [comment]
> It actually looks like `expectNoMessage` obtains the unwrapped version of the message received in `receiveN` above:

This sentence is wrong. `RealMessage(1,Actor[akka://BalancingSpec/user/balancingPool-3/$b#343717519])` means there was a response from node `1`. I thought `1` was some message `id` or counter. So the `expecteNoMsg` is not consuming a duplicate, it's consuming a second message that arrived unexpectedly.

**Comment by ignasi35** [comment]
Summing up: the BalancingPool and the test expect some context switching forcing the distribution of messages across the routees but that's not happening as much as the test expected.

BalancingPool relies on routees racing to the pool's shared inbox as soon as they are ready to process a new message. Therefore, we probably need to reconsider the test.

**Comment by ignasi35** [comment]
> The `RealMessage` wrapper _seems_ like a `TestKit` internal that we should perhaps unpack and doesn't have anything to do with this test?

It looks like `expectNoMessage` has behaved like that for a while. In testkit, the `receiveOne` and `receiveN` methods do unwrap `RealMessage`. Even the `expectMsg` overloads do the unwrapping (since they're implemented in terms of `receiveX`).

IMHO, it's not that bad that `expectNoMessage` returns `RealMessage(msg, sender)` since users get the extra information of what the sender was (it was useful for us in this case of BalancingSpec).

**Comment by raboof** [comment]
> since users get the extra information of what the sender was (it was useful for us in this case of BalancingSpec).

That is indeed useful!

**Comment by patriknw** [comment]
https://github.com/akka/akka/runs/3378509163?check_suite_focus=true#step:5:1599
